### PR TITLE
Patch lambda to allow imports

### PR DESCRIPTION
**Pull request by VenelinMartinov**
Should address https://github.com/pulumi/pulumi-aws/issues/2392

We regressed in our handling of imported lambdas in 5.28, when we picked up upstream https://github.com/hashicorp/terraform-provider-aws/releases/tag/v4.51.0 with https://github.com/hashicorp/terraform-provider-aws/pull/28963/files, which introduces the `ExactlyOneOf` constraints in the lambda resource.

This PR reverts the upstream `ExactlyOneOf` constraints and replaces them with the previously applied `ConflictsWith`. This in turn allows imports of lambdas via `pulumi import` and the `import` resource option to work properly.

I've added a GRPC test for a lambda imported via `pulumi import` as well as an integration test for a lambda imported via the `import` resource option and a test which checks that we still fail to create lambdas without any code-related properties (which should be the only case that now passes the validation step which previously didn't).

Note that this slightly worsens the behaviour in the case when none of the code-related properties are specified. Previously that'd tirgger a failure during `Check` and print a sensible error message but now we will fail during `Create` with the following much less legible error:

```
  aws:lambda:Function (testLambda):
    error: 1 error occurred:
        * creating Lambda Function (testLambda-e9e5e22): operation error Lambda: CreateFunction, https response error StatusCode: 400, RequestID: 88b5f9d7-42d6-4a3b-becc-96cdf954f596, api error ValidationException: 2 validation errors detected: Value '' at 'code.s3Key' failed to satisfy constraint: Member must have length greater than or equal to 1; Value '' at 'code.s3Bucket' failed to satisfy constraint: Member must have length greater than or equal to 3
```

I don't see a sensible way around it though and I think the change is still worthwhile.

Summary of effects:
| | <5.28 | 5.28-6.22 | This PR |
| :-------------  | :-------: | :----:| :-------:|
| pulumi import  | works? | doesn't work | works|
| import option   | works |doesn't work | works|
| no code lambda| errors during update?| errors during validation | errors during update |
| other lambdas| works| works  | works |

### CODE REVIEW

**Comment by VenelinMartinov** [comment]
Actually thinking a bit more on the bad error message, `Value '' at 'code.s3Key' failed to satisfy constraint:` must mean that none of the code related properties are specified, so we can probably replace the error message with something more sensible. I'll try that next.

File: examples/examples_nodejs_test.go [hunk 1]
```diff
@@ -565,3 +567,35 @@ func getAwsSession(t *testing.T) *session.Session {
...
+
+	// Check that we can reimport the lambda.
+	secondStack.SetConfig("lambda_name", lambdaName.Value.(string))
+	secondStack.SetConfig("runtime", "nodejs18.x")
+	secondStack.SetConfig("lambda_role", lambdaRole.Value.(string))
+	secondStack.Up()
+
+	// Check that we can change a property on the lambda
+	secondStack.SetConfig("runtime", "nodejs16.x")
+	secondStack.Up()
```
Comment (t0yv0) [inline examples/examples_nodejs_test.go:591]: There's no asserts here are we sure this is a strong test?

**Comment by t0yv0** [comment]
Tests are crucial here as the patches are a brittle mechanism an we are trying to move fast - so we need good reminders where it stops working. Would appreciate solutions without patches long-term e.g. lean into supported primitives if possible..

**Review by t0yv0 — APPROVED** [review]

File: examples/examples_nodejs_test.go (hunk 1 above)
Comment (VenelinMartinov) [reply examples/examples_nodejs_test.go:591]: The framework has the asserts by default: https://github.com/pulumi/providertest/blob/dc14d64784583753791733c6ac090d972d1ec746/pulumitest/up.go#L18

I've also verified that the test fails without the patch.

**Comment by VenelinMartinov** [comment]
>Can you create a bridge issue to support ExactlyOneOf in import?

The issue is not the bridge support for `ExactlyOneOf` - when we call `Read` on the lambda we get none of the `ExactlyOneOf` properties - instead we get a `sourceCodeHash`. That seems to be the way the code is represented after getting uploaded initially. I've pasted the GRPC logs from an imported lambda here: https://github.com/pulumi/pulumi-aws/pull/3371#issuecomment-1924506770

We (and TF) previously supported creating a lambda without any of the code-related propeties (filename, s3_bucket, image_uri) but since the addition of the ExactlyOneOf this no longer works. Creating lambdas without code-related properties is necessary for imports. And for imports via the resource option we don't even have `sourceCodeHash` in the inputs: https://github.com/pulumi/pulumi-aws/issues/2392#issuecomment-1939436045

I am not aware of any non-patch way of handling this, LMK if you disagree.

**Comment by t0yv0** [comment]
How does pure TF handle this import.

**Pull request merged**